# Coding One Hidden Layer

## Method 1: Specify everything explicitly

In this version, we will need to specify the input, parameters, and operations explicitly. Let's recall the architecture of the network we saw in the lectures:

<img src="Lecture 7-5.jpg" width=600 align="center">

Here we will reproduce the example calculation that we did by hand in **Lecture 7**. We first need to store the input values in separate variables. Here, we are considering only one *sample* and each sample has only *two* features. 

In [1]:
x_1 = 2 
x_2 = 3

We now need to specify the parameters (the *weights* and *bias*) for each neuron in each layer. We talked a little bit about how a network *learns* but we will go into more detail about that next week. In order to do the first *forward pass* of the network we need to *initialize the parameters*, that is, pick starting values for the weights and biases. How you do this initialization can have an impact on how well the network learns. For our purposes right now, we just want to have some numbers so we can do the calculation. 

### Layer 1

For layer 1, we have 2 neurons, and each of those neurons has 2 weights and 1 bias. We will use the notation *w = weight*, *b = bias*, *L = layer*, and *N = neuron* plus a number. So, `w1_N1_L1` is the first weight of the first neuron in layer 1, and so on.  And `b_N1_L1` is the bias for the first neuron of the first layer (we don't need `b1` and `b2` because each neuron only has a single bias), and so on.

In [2]:
w1_N1_L1 = 0
w2_N1_L1 = 1
b_N1_L1 = 2

w1_N2_L1 = 1
w2_N2_L1 = 1
b_N2_L1 = 1

We will now calculate the $z$ values for the 2 neurons in layer 1: 

In [3]:
z_N1_L1 = w1_N1_L1 * x_1 + w2_N1_L1 * x_2 + b_N1_L1
z_N2_L1 = w1_N2_L1 * x_1 + w2_N2_L1 * x_2 + b_N2_L1

print("z_N1_L1 = ", z_N1_L1)
print("z_N2_L1 = ", z_N2_L1)

z_N1_L1 =  5
z_N2_L1 =  6


We now need to apply our *activation function* to these `z` values. We are going to use the *sigmoid* function, as we did in the lecture, so we need to first create a function for that. 

In [4]:
import numpy as np

def sigmoid(z):
    a = 1 / (1 + np.exp(-z))
    return a

And then apply it:

In [5]:
a_N1_L1 = sigmoid(z_N1_L1)
a_N2_L1 = sigmoid(z_N2_L1)

print("The output of N1_L1 is: ", a_N1_L1)
print("The output of N2_L1 is: ", a_N2_L1)

The output of N1_L1 is:  0.9933071490757153
The output of N2_L1 is:  0.9975273768433653


The above output for each neuron in layer 1 is the same as what we did by hand in the lecture. (See the image at the beginning of this notebook.)

### Layer 2

For layer 2, our output layer, we have only 1 neuron. This neuron has 2 weights and 1 bias.

In [6]:
w1_N1_L2 = 2
w2_N1_L2 = 0
b_N1_L2 = 1

We will now calculate the $z$ value for this neuron, remembering that the inut values are no longer our training data but the output of the 2 neurons in layer 1: 

In [7]:
z_N1_L2 = w1_N1_L2 * a_N1_L1 + w2_N1_L2 * a_N2_L1 + b_N1_L2

print("z_N1_L2 = ", z_N1_L2)

z_N1_L2 =  2.9866142981514305


And then apply the sigmoid activation function:

In [8]:
a_N1_L2 = sigmoid(z_N1_L2)

print("The output of N1_L2 is: ", a_N1_L2)

The output of N1_L2 is:  0.9519657289209066


We can see that this is the same result as we had in the lecture. 

## Method 2: Using NumPy arrays 

It should be obvious that Method 1 would be pretty much impossible for any realistic network, for example, one that takes in hundreds of feature values, has 5 hidden layers with 25 neurons each, etc. However, it is a good learning exercise to go through implementing our simpler network in this manner. 

So, for this method, we will repeat the example we did for Method 1 but use arrays and array operations instead. This will make the code simpler, less error prone, and faster. It will also allow us to put many samples through a forward pass of the network at once. Here, sample 1 = `[2, 3]` and sample 2 = `[4, 5]`.

In [ ]:
X = np.array([[2, 3], [4, 5]])
W_L1 = np.array([[0, 1], [1, 1]])
b_L1 = np.array([2, 1]) 

In [ ]:
Z_L1 = np.dot(W_L1, X.T) + b_L1.reshape(-1, 1)
Z_L1

In [ ]:
A_L1 = sigmoid(Z_L1)
A_L1

We now see that the final output is an array with 4 numbers. The first row correspondes to the output of the neurons in layer 1 for when the input = `[2, 3]`. That is, when our input is `[2, 3]`, the first neuron of layer 1 will output `0.99330715` and the second neuron of layer 1 will output `0.99752738`. Similarly, when our input is `[4, 5]`, the first neuron of layer 1 will output `0.99908895` and the second neuron of layer 1 will output `0.9999546`.

The second layer has one neuron with 2 weights and 1 bias:

In [ ]:
W_L2 = np.array([2, 0])
b_L2 = np.array(1) 

This layer now takes the output of layer 1 as its input, calculates a value for $z$ and applies the sigmoid activation function. 

In [ ]:
Z_L2 = np.dot(W_L2, A_L1) + b_L2.reshape(-1, 1)
Z_L2

In [ ]:
A_L2 = sigmoid(Z_L2)
A_L2

Thus, when our input is `[3, 4]`, our neural network will output 0.95196573; or, you could say that for sample `[3, 4]`, our neural network will predict 0.95196573. And when our input is `[4, 5]`, our neural network will output 0.95249174; or, you could say that for sample `[4, 5]`, our neural network will predict 0.95249174. 

## Method 3: Creating reusable objects

As we did last week, we will now create our neural network using Python **classes**. Again, just focus on how this simplifies what we are trying to accomplish. 

In the code below, we are creating a recipe for a network with a single hidden layer. **Please note that we have not yet created an actual network.** This is similar to the difference between having a recipe for baking a loaf of bread and actually baking a loaf of bread. 

The **__init__** function will initialize all the weights and biases for all the neurons in each layer. The **feedforward** function then takes the input you provide to the network and does all of the array operations necessary for the network to produce output.

In [ ]:
# Inspired by code written by Victor Zhou
# License for reuse: https://github.com/vzhou842/neural-network-from-scratch/blob/master/LICENSE

import numpy as np # included again here for completeness; normally this should only occur once in your notebook

class OurNeuralNetwork:
  '''
  A neural network with:
    - 2 inputs
    - a hidden layer with 2 neurons (h1, h2)
    - an output layer with 1 neuron (o1)
  '''
  def __init__(self, w_L1, w_L2, b_L1, b_L2):
    # initialize the weights and biases for each layer
    self.w1 = w_L1 
    self.b1 = b_L1.reshape(-1, 1)
    self.w2 = w_L2
    self.b2 = b_L2.reshape(-1, 1)
  
  def activation(self, z):
    a = 1 / (1 + np.exp(-z))
    return a

  def feedforward(self, input):
    Z1 = np.dot(self.w1, input.T) + self.b1
    A1 = self.activation(Z1)
    Z2 = np.dot(self.w2, A1) + self.b2
    A2 = self.activation(Z2)

    out = A2

    return out

Let's define some values for the weights and bias.

In [ ]:
w_L1 = np.array([[0, 1], [1, 1]]) 
b_L1 = np.array([2, 1])

w_L2 = np.array([[2, 0]])
b_L2 = np.array([1])

Now let's use the recipe to create a single (specific) neuron. (That is, we are now using the recipe to bake a particular loaf of bread.)

In [ ]:
network = OurNeuralNetwork(w_L1, w_L2, b_L1, b_L2)

The line of code above creates a neuron with specific values for the weights and bias. It also already has a function that we can use to carry out the appropriate mathematical operations of our neuron to produce output whenever we supply values for the input. 

In [ ]:
x = np.array([[2, 3], [4, 5]])

network.feedforward(x) 